monai/apps/generation/maisi/networks/autoencoderkl_maisi.py

```diff
@@ -0,0 +1,901 @@
+        if outputs[0].size(non_dim_split + 2) // splits[0].size(non_dim_split + 2) == 2:
+            split_size_out *= 2
+            padding_s *= 2
+        elif splits[0].size(non_dim_split + 2) // outputs[0].size(non_dim_split + 2) == 2:
+            split_size_out //= 2
+            padding_s //= 2
+
+        slices = [slice(None)] * 5
+        for i in range(num_splits):
+            slices[self.dim_split + 2] = (
+                slice(None, split_size_out) if i == 0 else slice(padding_s, padding_s + split_size_out)
+            )
+            outputs[i] = outputs[i][tuple(slices)]
+
+        if self.debug:
+            for i in range(num_splits):
+                print(f"outputs after {i + 1}/{len(outputs)}:", outputs[i].size())
+
+        if max(outputs[0].size()) < 500:
+            x = torch.cat(outputs, dim=self.dim_split + 2)
+        else:
+            x = outputs[0].clone().to("cpu", non_blocking=True)
+            outputs[0] = 0
+            _empty_cuda_cache()
+            for k in range(len(outputs) - 1):
+                x = torch.cat((x, outputs[k + 1].cpu()), dim=self.dim_split + 2)
+                outputs[k + 1] = 0
+                _empty_cuda_cache()
+                gc.collect()
+                if self.debug:
+                    print(f"MaisiConvolution cat: {k + 1}/{len(outputs) - 1}.")
+            x = x.to("cuda", non_blocking=True)
+
+        del outputs
+        _empty_cuda_cache()
+
+        return x
+
+
+class MaisiUpsample(nn.Module):
+    """
+    Convolution-based upsampling layer.
+
+    Args:
+        spatial_dims: Number of spatial dimensions (1D, 2D, 3D).
+        in_channels: Number of input channels to the layer.
+        use_convtranspose: If True, use ConvTranspose to upsample feature maps in decoder.
+        num_splits: Number of splits for the input tensor.
+        dim_split: Dimension of splitting for the input tensor.
+        debug: Whether to print debug information.
+    """
+
+    def __init__(
+        self, spatial_dims: int, in_channels: int, use_convtranspose: bool, num_splits: int, dim_split: int, debug: bool
+    ) -> None:
+        super().__init__()
+        self.conv = MaisiConvolution(
+            spatial_dims=spatial_dims,
+            in_channels=in_channels,
+            out_channels=in_channels,
+            strides=2 if use_convtranspose else 1,
+            kernel_size=3,
+            padding=1,
+            conv_only=True,
+            is_transposed=use_convtranspose,
+            num_splits=num_splits,
+            dim_split=dim_split,
+            debug=debug,
+        )
+        self.use_convtranspose = use_convtranspose
+
+    def forward(self, x: torch.Tensor) -> torch.Tensor:
+        if self.use_convtranspose:
+            x = self.conv(x)
+
+            if isinstance(x, torch.Tensor):
+                return x
+            return torch.tensor(x)
+
+        x = F.interpolate(x, scale_factor=2.0, mode="trilinear")
```

Comment:
`scale_factor` must be fixed for upsampling. And we cannot change `mode` to other values when generating image with size 512 x 512 x 512 since PyTorch would return precision error.